README.md
<img src="logo/CTestLogo-200.png" alt="CTest logo" align="right" />

# CTest: an online system to track COVID-19 test results

This is an online tracking system for COVID-19 tests. CTest provides a browser-based online status update for testees using personalized QR-codes (and web links). CTest does not require installation of any apps or logins for users. Instead, CTest uses cryptographically secure tracking IDs and does not use any personal data.
The primary aim of this approach was to reduce the burden of clinical staff in the COVID-19 crisis and to allow efficient and almost instantaneous communication of the results to testees. Speedy communication is of the essence in the current crisis, as it virus carriers can be infectious before first symptoms arise.


## Table of Contents
   
  * [Getting Started](#getting-started)
     * [Prerequisites](#prerequisites)
     * [Initialisation and Startup](#initialisation-and-startup)
  * [Configuration](#configuration)
     * [General Settings](#general-settings)
     * [Server Settings for CTest](#server-settings-for-ctest)
     * [Order numbers &amp; Import](#order-numbers--import)
     * [Backup](#backup)
     * [Branding](#branding)
     * [CTest server setups with Apache as proxy](#ctest-server-setups-with-apache-as-proxy)
     * [Protection from Brute-Force Attacks](#protection-from-brute-force-attacks)            
  * [Using CTest](#using-ctest)
     * [Administration](#administration)
     * [User](#user)
     * [If something went wrong](#if-something-went-wrong)
     * [Testee](#testee)
     * [App integration](#app-integration)
  * [Build](#build)
  * [Built With](#built-with)
  * [Authors](#authors)
  * [License](#license)
  * [Acknowledgments](#acknowledgments)
  * [Cite](#cite)


## Getting Started

These instructions will explain you how to setup a local CTest instance on your computer for testing purposes. 

Download a release version of CTest: 

* [Version 0.3.2](https://github.com/sysbio-bioinf/CTest/releases/download/0.3.2/ctest-0.3.2.jar)

For testing purposes also download the [`keystore.jks`](https://github.com/sysbio-bioinf/CTest/raw/master/keystore.jks)
or [disable SSL/https](#server-settings-for-ctest) usage.


### Prerequisites

CTest has been tested with **Java 8, 11 and 14.**

CTest should be run with Transport Layer Security (TLS).
Typically, you will run CTest behind an existing web server (apache, nginx,...) that already provides TLS.

For testing purposes the keystore `keystore.jks` in the project repository has been generated as follows:

``` shell
$ keytool -genkey -keyalg RSA -alias selfsigned -keystore keystore.jks -storepass password -validity 365 -keysize 2048
```
*A self-signed certificate has the disadvantage that the users are warned by their web browsers that their connection is not secure. Hence, this should only be used for test setups.*

### Initialisation and Startup

CTest is initialised via `java -jar ctest-<version>.jar init` which allows the following parameters:

``` shell
$ java -jar ctest-<version>.jar init -h
Initialise the CTest instance.

  -a, --admin NAME           admin_user       Name of the admin user
  -p, --password SECRET      secret_password  Admins password
  -d, --data-base-name NAME  ctest.db         Name of the database. CTest will not override a existing database file.
  -h, --help
```

`--admin`
 : login name of the admin account

`--password`
 : password of the admin account

`--data-base-name`
 : filename of the CTest database to create (existing databases are not overridden)

The `init` call will create the CTest database with the default or specified administrator user and a default configuration file.
The configuration file should be adjusted to the concrete usage scenario.

After the initialisation CTest can be started as follows:

``` shell
$ java -jar ctest-<version>.jar run
CTest started - Server listening on:
http://localhost:8000
https://localhost:8443
```

The default configuration will start CTest only on the localhost -- to be able to access the server remotely the server settings need to be adjusted.

To use different configuration files, you can specify the configuration to use as follows:

``` shell
$ java -jar ctest-<version>.jar run -c another-ctest.conf
```

## Configuration

The CTest configuration file allows customisation for different setup scenarios.
The configuration file contains a Clojure map to specify the various settings.
Consider the following short excerpt from the default configuration:

``` clojure
{:server-config
 {:host "localhost",
  :port 8000},

 :data-base-name "ctest.db"
 ... }
```

Each map is started by `{` and ended by `}`.
Settings are specified as pairs of keyword and value, e.g. keywords such as `:port` followed by a value such as `8000`.
Keywords also identify configuration for subcomponents of CTest, e.g. the server configuration `:server-config`.


### General Settings

CTest offers the following general settings for customisation:

``` clojure
{ ...
 :data-base-name "/installpath/ctest/ctest.db"
 
 :log-level :info,
 :log-file "ctest.log",

 ... }
```

The effect of these settings is described in the following.

`:data-base-name`
 : path to the database that CTest uses to store all user account and tracking data.

`:log-level`
 : specifies how much information is written into the log file. Possible values: `:trace`, `:debug`, `:info`, `:warn`, `:error`, `:fatal`

`:log-file`
 : path to the log file.


### Server Settings for CTest

CTest ships with a built-in web server ([Jetty](http://www.eclipse.org/jetty/)).

The following listing shows the relevant server settings.
Other settings together with a short description can be found in the [ring library documentation](https://ring-clojure.github.io/ring/ring.adapter.jetty.html).

``` clojure
{:server-config
 {:host "localhost",
  :port 8000,
  :server-root "",

  :proxy-url nil,

  :ssl? true,
  :ssl-port 8443,
  :keystore "keystore.jks",
  :key-password "password"},
 ... }
```

The server settings are located in the `:server-config`.
The following properties are used:

`:host`
 : specifies the network interface to which the web server binds.
   The host is also used for the generation of links to CTest pages.
   Use `"localhost"` for access from the local computer.
   Use the full qunatified domain name or IP of the network interface that connects to the internet for remote access.

`:port`
 : specifies the port on which the web server will listen.

`:server-root`
 : specifies the prefix path under which the CTest server pages are available, e.g. `"ctest"` when CTest is run at `your.domain.tld/ctest`

`:proxy-url`
 : When CTest is used behind a proxy, you need to specify the URL under which CTest is accessible at the proxy.

`:ssl?`
 : enables or disables transport encryption (aka "https").

`:ssl-port`
 : specifies the port used for encrypted communication.

`:keystore`
 : specified the filename of the keystore that contains the certificate for the transport encryption.

`:key-password`
 : specifies the password needed to access the keystore (if any).


### Order numbers & Import

The format of **order numbers** needs to be specified by a regular expression.
In some scenarios the uniqueness of order numbers can only be guaranteed for the same day.
Therefore, you can specify `:append-date? true` such that CTest constructs order numbers 
from the given numbers `XXXXXXXX` by appending the current date resulting in `XXXXXXXX-YYYYMMDD`.
If this is not required, specify `:append-date? false`.
A configuration that appends dates may look as follows:
``` clojure
:order-numbers {:input-format "[0-9]{8,8}"
                :append-date? true
                :allow-tracking-number-usage? false
                :new-patient-hint "Order Number"
                :new-patient-note "Note: Order Number must be an eight digit number."
                :find-patient-hint "XXXXXXXX-YYYYMMDD"
                :find-patient-note "Note: Order Number must be an eight digit number combined with eight digit date separated by a hyphen."}
```
Hints about valid order numbers for new patients and patient identifiers to find patients in the web front end 
can be configured using ```new-patient-hint``` and ```new-patient-note``` as well as ```find-patient-hint``` and ```find-patient-note```.
In a scenario without order numbers `:allow-tracking-number-usage? true` can be specified 
to activate an additional page in the menu called `New Tracking` that allows
the creation of patient records without an order number.
In this case the tracking number is used as order number. 

You have to specify the format of the CSV files and the location (`:path`) where CTest can find them:
``` clojure
:import {:path "/path/to/import/folder"
         :order-number-column "auftragsn"
         :date-column "abnahme"
         :date-format "dd.MM.yyyy"
         :result-column "ergebnis"
         :negative-result "ungr"
         :column-separator ";"
         :antibody-results? false
         :reactive-result "reak"
         :non-reactive-result "nreak"}
```

CTest will keep watching the specified `:path` for new csv files.
When no dates are appended to the order numbers, two columns are required:

1. the column containing the order number (`:order-number-column`).
2. the column containing the result of the test (`:result-column`).

In case dates are appended to the order numbers, a third column (`:date-column`) is needed containing
the date of sample collection (the same date the patient was entered into CTest).
Furthermore, the date format needs to be specified (`:date-format`) 
with placeholders `d` for day digit, `M` for month digit and `y` for year digit (see [Javadocs](https://docs.oracle.com/javase/8/docs/api/index.html?overview-summary.html)).

You have to specify the correct `:column-separator` that is used in the CSV files.

By setting `:antibody-results? true`, CTest is able to import and show antibody test results, as well.
Two results are supported for antibody test: `reactive` and `non-reactive`.
You need to specify how these two results are encoded in the result column.

### Backup

CTest performs a backup on startup and with the default configuration 
hourly backups every half an hour.
You can adjust the settings in the configuration file:

``` clojure
:backup {:path "/safestorage/ctest/backup"
         :start-minute 30
         :interval 60}
```

`:path`
 : path to the directory where the project stores backups from the database.
 
`interval` 
 : interval between two consecutive backups.
 
`:start-minute`
 : the minute of an hour when the first backup shall be performed (relative to the start of CTest).

If the configuration does not contain a `:backup` entry, CTest will not do any backup.

See [Restore a backup](#restore-a-backup) for instructions to restore a database backup. 


### Branding

You can customize your instance of CTest by modifying the following parameters in the configuration file ```ctest.conf```:

``` clojure
:branding {:path "storage/branding"
           :page-logo "logo.png"
           :patient-document-logo "pat_logo.png"
           :page-title "Corona Virus Information",
           :page-title-link "http://your.hospital.org/corona"
           :institute {:name "Clinical Chemistry"
                       :url "http://clinchem.your.hospital.org"}}
```
First, specify the ```path```, where your logo images are stored. 
The parameters ```page-logo```and ```patient-document-logo``` determine the filenames of the logo 
that is shown on top of the page and the one that is included in the patient printouts. 
```page title```and ```page title link``` can be used as header and link to your organization unit.
When antibody results are displayed, you can configure the institute name (`:institute :name`) 
and an url to additional information (`:institute :url`).


### CTest server setups with Apache as proxy

This section explains a fully featured setup on a Linux server with automatic startup of CTest when the server is rebooted.
There are two setup scenarios for CTest:

1. CTest standalone
2. CTest behind proxy

*A similar setup is possible on windows servers as well, though the means to achieve it are different.*

#### Automatic Startup of CTest

CTest can be installed as a daemon on Linux via using the init script [ctest](https://raw.githubusercontent.com/sysbio-bioinf/ctest/master/tools/ctest) (root privileges on the server are required).
Modify the settings in the init script appropriately and copy it to `/etc/init.d/`.
The settings section of the init script looks as follows:

``` shell
## CTest settings

VERSION="<version>"
INSTALL_DIR="/home/youruser/ctest"
CONFIG="ctest.conf"
OPTS=""

## Linux user and group for CTest process
RUN_AS_USER="ctest"
RUN_AS_GROUP="ctest"

## memory settings for the JVM
MIN_MEMORY="256M"
MAX_MEMORY="4096M"
```

Set the install directory `INSTALL_DIR` to the one where you copied the CTest jar file and initialised the database.
If you choose a different configuration name, you will have to adjust the variable `CONFIG` appropriately.
It is recommended to create a user account dedicated to run CTest.
In the example above a user account `ctest` with group `ctest` is used.
The minimal and maximal memory that is used by the JVM can be adjusted to your usage scenario.

After copying the init script `ctest` to `/etc/init.d/ctest` you can configure the automatic startup on server boot
via the operating system tools, e.g.  `rcconf`.
Manual start, stop and restart is possible via the following:

``` shell
/etc/init.d/ctest start
/etc/init.d/ctest stop
/etc/init.d/ctest restart
```
or
``` shell
service ctest start
service ctest stop
service ctest restart
```


##### CTest Standalone

For the standalone scenario only the init script needs to be setup as described in the Section [Automatic Startup](automatic-startup).
The `:server-config` in the configuration file should set the port to `80`, the ssl port to `443` and the domain name of your server as host.

``` clojure
{:server-config
 {:host "your.domain.tld",
  :port 80,
  :ssl? true,
  :ssl-port 443,
  ... },
 ... }
```


##### CTest behind Proxy

This section describes the CTest setup using [Apache](https://httpd.apache.org/) as proxy.
In the proxy scenario the transport encryption is provided by Apache such that transport encryption is disabled in CTest.
Therefore, the init script setup from Section [Automatic Startup](automatic-startup) should be performed in advance.

We describe two possibilities to use Apache as a proxy for CTest:

1. CTest as a subdomain of your domain (virtual host).
2. CTest as a subdirectory.


##### CTest as Subdomain

To setup CTest to use a subdomain of your domain (virtual host),
you need to have a virtual host definition in `/etc/apache2/sites-available/ctest` like the following:

``` apache
ProxyRequests Off
ProxyPreserveHost On

<VirtualHost *:80>
  ServerName ctest.your.domain.tld
  Redirect permanent / https://ctest.your.domain.tld/
</VirtualHost>


<VirtualHost *:443>
  ServerName ctest.your.domain.tld

  SSLEngine On
  SSLProxyEngine On
  SSLProxyVerify none
  SSLCertificateFile /path/to/certificate.crt
  SSLCertificateKeyFile /path/to/keyfile.key

  ProxyPass / http://localhost:CTEST-PORT/
  ProxyPassReverse / http://localhost:CTEST-PORT/
  RequestHeader set X-Forwarded-Proto "https"
  ErrorLog /var/log/apache2/ctest.log
  CustomLog /var/log/apache2/ctest.log common
</VirtualHost>
```

This sets up a a virtual host `ctest.your.domain.tld` on your domain `your.domain.tld` using transport encryption.
The virtual host is forwarded to the chosen port `CTEST-PORT` on `localhost` which CTest listens to.
The paths to the certificate file `/path/to/certificate.crt` and the certificate key file `/path/to/keyfile.key` 
must be filled in according to your server configuration.
The `:server-config` of the CTest configuration has to look like:

``` clojure
{:server-config
 {:host "localhost",
  :port CTEST-PORT,
  :server-root "",

  :proxy-url "ctest.your.domain.tld",

  :ssl? false},
 ... }
```

##### CTest as Subdirectory

To setup CTest to use a subdirectory of your domain, you need to add the following to your existing virtual host:

``` apache
<Location /ctest>
  ProxyPass        http://localhost:CTEST-PORT/ctest
  ProxyPassReverse http://localhost:CTEST-PORT/ctest
</Location>
```

You have to fill in `CTEST-PORT` with the port you specified in your CTest configuration file.
You also need to specify `ProxyRequests Off` and `ProxyPreserveHost On` before the existing virtual host, e.g.

``` apache
ProxyRequests Off
ProxyPreserveHost On

<VirtualHost *:443>
  ServerName your.domain.tld
  
  ...
  
  <Location /ctest>
    ProxyPass        http://localhost:CTEST-PORT/ctest
    ProxyPassReverse http://localhost:CTEST-PORT/ctest
  </Location>
</VirtualHost>
```

This sets up a directory `ctest` on your domain `your.domain.tld` such that CTest is accessible via `your.domain.tld/ctest`.
The directory `ctest` of the virtual host is forwarded to the chosen port `CTEST-PORT` on `localhost` which CTest listens to.
The `:server-config` of the CTest configuration has to look like:

``` clojure
{:server-config
 {:host "localhost",
  :port CTEST-PORT,
  :server-root "ctest",

  :proxy-url "your.domain.tld",

  :ssl? false},
 ... }
```

### Protection from Brute-Force Attacks

[Fail2Ban](https://github.com/fail2ban/fail2ban) can be used as a protection from
brute-force attacks on logins and tracking URLs.
The following instructions describe the setup of Fail2Ban for CTest on a Linux server.
The configuration consist of at least 3 files:
 * [`filter.d/ctest.conf`](tools/fail2ban/filter.d/ctest.conf): 
   defines the filter to detect failed login attempts
 * [`filter.d/ctest-tracking.conf`](tools/fail2ban/filter.d/ctest-trackig.conf): 
   defines the filter to detect tracking URL guessing/enumeration attempts
 * [`jail.local`](tools/fail2ban/jail.local):
   defines the rules for banning IPs due to brute-force attacks on logins or tracking URLs
  
A fourth file is required, when you are using a reverse proxy that is running on a different server
than CTest:
 * [`action.d/iptables-reverse-proxy.conf`](tools/fail2ban/action.d/iptables-reverse-proxy.conf):
   defines the actions to ban and unban the source IPs (`X-Forwarded-For:`)
   instead of the proxy IP
   
Apart from `jail.local`, you can just copy the files to the configuration directory of Fail2Ban
`/etc/fail2ban`. If you already use Fail2Ban, you will have to merge our `jail.local` with your current `jail.local`.

There are a few parameters that you can adjust to your needs in `jail.local`, 
e.g. for preventing login brute-force attacks:
```
[ctest]

enabled  = true
#ignoreip = whitelisted ips
port     = 80
protocol = tcp
filter   = ctest
logpath  = /opt/ctest/ctest.log
maxretry = 3
bantime  = 180
findtime = 180
banaction = iptables-reverse-proxy[name = ctest, port = http, protocol = tcp]
```


CHANGELOG.md
# Version 0.3.2

* Adds support to import and show antibody test results (`:antibody-results? true`).

# Version 0.3.1

* Adds (optional) tracking without order numbers (`:allow-tracking-number-usage? true`).

# Version 0.3.0

* Upgrade to Clojure 1.10.1.

* Upgrade of all dependencies.

* Tested with Java 8, Java 11 and Java 14.

* Removes configuration parameter `:forwarded?`.

* Switched `import` task to batched insert.

* Switched CSV import to batched update.

* Default configuration sets `:ssl? false`.

* Fixes shutdown logic in import loop.

* New `upgrade` task to migrate from `0.1.3` and `0.2.x` to `0.3.x`. 

# Version 0.2.2

* Fixes a bug in user management dialog.

* Removes unnecessary JavaScript includes.


# Version 0.2.1

* Adds an *app access*. By appending `?app=true` to a tracking URL, CTest will only return "negative" or "in progress" instead of the complete HTML.

* The backup daemon configuration allows specification of the backup schedule and changed from `:backup-path` to the following:
  ```clojure
  :backup {:path "backup"
           :start-minute 30
           :interval 60}
  ```
 
* The import daemon configuration allows the specification of the `:column-separator`:
  ```clojure
  :import {...
           :column-separator ";"
           ...}
  ```
 
 * Reports can now be filtered by report `type` and `context` by querying `/reports/list?type=error&context=CSV`.
   

src/ctest/config.clj
; Copyright (c) Fabian Schneider and Gunnar Völkel. All rights reserved.
; The use and distribution terms for this software are covered by the
; Eclipse Public License 2.0 (http://www.eclipse.org/legal/epl-v20.html)
; which can be found in the file epl-v2.0.txt at the root of this distribution.
; By using this software in any fashion, you are agreeing to be bound by
; the terms of this license.
; You must not remove this notice, or any other, from this software.

(ns ctest.config
  (:require
    [clojure.string :as str]
    [clojure.pprint :refer [pprint]]
    [clojure.stacktrace :as st]
    [clojure.java.io :as io]
    [ctest.common :as common]
    [clojure.tools.logging :as log])
  (:import
    com.mchange.v2.c3p0.ComboPooledDataSource
    (org.apache.log4j PropertyConfigurator)))



(defonce ^:private ctest-config
  (atom
    {:log-level :info
     :log-file "ctest.log"
     :data-base-name "ctest.db"
     :qrcode-path "storage/qrcodes"
     :branding {:path "storage/branding"
                :page-logo "logo.png"
                :patient-document-logo "logo.png"
                :page-title "Corona Virus Information",
                :page-title-link "http://your.hospital.org/corona"
                :institute {:name "Clinical Chemistry"
                            :url "http://clinchem.your.hospital.org"}}
     :order-numbers {:input-format "[0-9]{8,8}"
                     :allow-tracking-number-usage? false
                     :append-date? true
                     :new-patient-hint "Order Number"
                     :new-patient-note "Note: Order Number must be an eight digit number."
                     :find-patient-hint "XXXXXXXX-YYYYMMDD"
                     :find-patient-note "Note: Order Number must be an eight digit number combined with eight digit date separated by a hyphen."}
     :import ^:replace {:path "import"
                        :order-number-column "auftragsn"
                        :date-column "abnahme"
                        :date-format "dd.MM.yyyy"
                        :result-column "ergebnis"
                        :negative-result "ungr"
                        :antibody-results? false
                        :reactive-result "reak"
                        :non-reactive-result "nreak"
                        :column-separator ";"}
     :backup {:path "backup"
              :start-minute 30
              :interval 60}
     :server-config ^:replace {:port 8000
                               :host "localhost"
                               :ssl? false
                               :ssl-port 8443
                               :server-root ""
                               :proxy-url nil
                               :keystore "keystore.jks"
                               :key-password "password"}}))


(defn write-config-file
  [filename, options]
  (let [config (merge @ctest-config options)]
    (spit filename
      (with-open [w (java.io.StringWriter.)]
        (pprint config w)
        (str w)))))


(defn server-config
  []
  (:server-config @ctest-config))


(defn page-title
  []
  (get-in @ctest-config [:branding, :page-title]))


(defn page-title-link
  []
  (get-in @ctest-config [:branding, :page-title-link]))


(defn develop?
  []
  (:develop? @ctest-config))


(defn tracking-server-domain
  []
  (let [{:keys [proxy-url, host, port, ssl?, ssl-port]} (server-config)]
    (or
      proxy-url
      (if ssl?
        (cond-> (str "https://" host) (not= ssl-port 443) (str ":" ssl-port))
        (cond-> (str "http://" host) (not= port 80) (str ":" port))))))


(defn server-root
  []
  (or (:server-root (server-config)) ""))


(defn server-location
  ([path]
   (server-location path, false))
  ([^String path, always-prefix-slash?]
   (let [{:keys [^String server-root]} (server-config)]
     (if (str/blank? server-root)
       path
       (let [slash? (.startsWith path "/")]
         (str
           (when (or slash? always-prefix-slash?) "/")
           server-root
           (when-not (or (str/blank? path) slash?) "/")
           path))))))


(defn trim-slashes
  [^String s]
  (when s
    (let [b (if (.startsWith s "/") 1 0)
          n (.length s)
          e (if (.endsWith s "/") (dec n) n)]
      (.substring s b e))))


(defn normalize-path
  [^String path]
  (when path
    (cond-> path
      (and (not (.endsWith path "/")) (not (str/blank? path)))
      (str "/"))))



(defn deep-merge
  "Merges the `source` map into the `sink` map such that nested map values are merged as well."
  [sink, source]
  (persistent!
    (reduce-kv
      (fn [res-map, k, source-value]
        (let [sink-value (get sink k)]
          (assoc! res-map k
            (cond
              (-> sink-value meta :replace)
              source-value,
              (and (map? sink-value) (map? source-value))
              (deep-merge sink-value, source-value)
              :else
              source-value))))
      (transient sink)
      source)))


(defn matches-pattern?
  [pattern, text]
  (boolean
    (when (string? text)
      (re-matches pattern text))))


(defn build-predicates
  [{:keys [input-format, append-date? date-format] :as order-numbers-config}]
  (let [input-regex (re-pattern input-format)
        order-number-regex (if append-date?
                             (re-pattern (str input-format "-[0-9]{8,8}"))
                             input-regex)]
    (assoc order-numbers-config
      :valid-order-number-input? (partial matches-pattern? input-regex)
      :valid-order-number? (partial matches-pattern? order-number-regex))))


(defn to-char
  [x]
  (cond
    (string? x) (first x)
    (char? x) x
    :else (char x)))


(defn update-config
  [config]
  (swap! ctest-config deep-merge
    (-> config
      (update-in [:server-config, :server-root] trim-slashes)
      (update-in [:qrcode-path] normalize-path)
      (update-in [:branding, :path] normalize-path)
      (update-in [:import, :path] normalize-path)
      (update-in [:import, :column-separator] to-char)
      (update-in [:backup, :path] normalize-path)
      (update-in [:order-numbers] build-predicates))))



(defn check-import-settings
  [{{:keys [path
            column-separator
            order-number-column
            result-column
            negative-result
            antibody-results?
            reactive-result
            non-reactive-result
            date-column
            date-format]} :import,
    {:keys [append-date?]} :order-numbers}]
  (cond-> []
    (str/blank? path)
    (conj "You did not specify a path for the CSV import.")

    (not (common/file-exists? path))
    (conj (format
            "You specified an import path \"%s\" for the CSV import which does not exist."
            path))

    (and column-separator (not (or (string? column-separator) (char? column-separator))))
    (conj (format
            "You specified an invalid :column-separator \"%s\" for the CSV import (valid example: \";\")."
            column-separator))

    (str/blank? order-number-column)
    (conj "You did not specify an :order-number-column for the CSV import.")

    (str/blank? result-column)
    (conj "You did not specify a :result-column for the CSV import.")

    (str/blank? negative-result)
    (conj "You did not specify the :negative-result value for the CSV import.")

    (and antibody-results? (str/blank? reactive-result))
    (conj "You specified to support antibody results (:antibody-results? true) but did not specify the :reactive-result value for the CSV import.")

    (and antibody-results? (str/blank? non-reactive-result))
    (conj "You specified to support antibody results (:antibody-results? true) but did not specify the :non-reactive-result value for the CSV import.")

    (and append-date? (str/blank? date-column))
    (conj "You did specify to use order numbers with appended date but did not specify a :date-column for CSV import.")

    (and append-date? (str/blank? date-format))
    (conj "You did specify to use order numbers with appended date but did not specify a :date-format for CSV import.")))


(defn check-order-numbers
  [{{:keys [input-format]} :order-numbers}]
  (cond-> []
    (str/blank? input-format)
    (conj "You did not specify an input format for the order numbers.")))


(defn check-config
  []
  (let [config @ctest-config]
    (not-empty
      (persistent!
        (reduce
          (fn [error-vec, check-fn]
            (if-let [errors (check-fn config)]
              (if (sequential? errors)
                (reduce conj! error-vec errors)
                (conj! error-vec errors))
              error-vec))
          (transient [])
          [check-import-settings
           check-order-numbers])))))


(defn database-config
  [db-filename]
  {:classname "org.sqlite.JDBC"
   :subprotocol "sqlite"
   :subname db-filename})


(def ^:private ^:once db-spec
  (atom (database-config "ctest.db")))


(defn pool
  [spec]
  (let [cpds (doto (ComboPooledDataSource.)
               (.setDriverClass (:classname spec))
               (.setJdbcUrl (str "jdbc:" (:subprotocol spec) ":" (:subname spec)))
               (.setUser (:user spec))
               (.setPassword (:password spec))
               ;; expire excess connections after 30 minutes of inactivity:
               (.setMaxIdleTimeExcessConnections (* 30 60))
               ;; expire connections after 3 hours of inactivity:
               (.setMaxIdleTime (* 3 60 60))
               ;; only one connection for SQLite
               (.setMinPoolSize 1)
               (.setMaxPoolSize 1))]
    {:datasource cpds}))


(def db-connection-pool (atom nil))


(defn db-connection
  []
  (if-let [db-conn @db-connection-pool]
    db-conn
    (swap! db-connection-pool #(if (nil? %) (pool @db-spec) %))))


(defn update-db-name [name]
  (swap! db-spec assoc :subname name)
  (reset! db-connection-pool nil))


(defn qrcode-path
  []
  (get-in @ctest-config [:qrcode-path]))


(defn branding-path
  []
  (get-in @ctest-config [:branding, :path]))


(defn page-logo
  []
  (get-in @ctest-config [:branding, :page-logo]))


(defn patient-document-logo
  []
  (get-in @ctest-config [:branding, :patient-document-logo]))


(defn branding-logo-path
  []
  (str (io/file (branding-path) (page-logo))))


(defn import-path []
  (get-in @ctest-config [:import, :path]))

(defn backup-path []
  (get-in @ctest-config [:backup, :path]))

(defn backup-config
  []
  (:backup @ctest-config))


(derive :role/admin :role/user)
(derive :role/admin :role/reporter)
(derive :role/configadmin :role/admin)

(defn set-daemon
  [daemon]
  (swap! ctest-config assoc :daemon daemon))

(defn daemon
  []
  (:daemon @ctest-config))

(defn set-backup-service
  [service]
  (swap! ctest-config assoc :backup-service service))

(defn backup-service
  []
  (:backup-service @ctest-config))


(defn order-number-input-format
  []
  (get-in @ctest-config [:order-numbers, :input-format]))


(defn order-number-append-date?
  []
  (get-in @ctest-config [:order-numbers, :append-date?]))


(defn order-number-date-format
  []
  (get-in @ctest-config [:order-numbers, :date-format]))


(defn allow-tracking-number-usage?
  []
  (get-in @ctest-config [:order-numbers, :allow-tracking-number-usage?]))


(defn order-numbers-config
  []
  (:order-numbers @ctest-config))


(defn institute-config
  []
  (get-in @ctest-config [:branding, :institute]))


(defn import-config
  []
  (:import @ctest-config))


(defn import-negative-result
  []
  (get-in @ctest-config [:import, :negative-result]))

(defn supports-antibody-results?
  []
  (get-in @ctest-config [:import, :antibody-results?]))

(defn valid-order-number-input-fn
  []
  (get-in @ctest-config [:order-numbers, :valid-order-number-input?]))


(defn valid-order-number-fn
  []
  (get-in @ctest-config [:order-numbers, :valid-order-number?]))


(defn import-csv-columns
  []
  (let [{:keys [order-number-column, date-column, result-column]} (import-config)
        column-names (cond-> [order-number-column, result-column]
                       (order-number-append-date?)
                       (conj date-column))]
    (mapv #(-> % str/trim str/lower-case keyword) column-names)))


(def ^:const db-encoding-in-progress "in progress")
(def ^:const db-encoding-negative "negative")
(def ^:const db-encoding-reactive "reactive")
(def ^:const db-encoding-non-reactive "non-reactive")

(defn valid-status-encoding?
  [x]
  (contains?
    #{db-encoding-negative, db-encoding-reactive, db-encoding-non-reactive, db-encoding-in-progress}
    x))

(defn negative-status?
  [s]
  (= s db-encoding-negative))

(defn reactive-status?
  [s]
  (= s db-encoding-reactive))

(defn non-reactive-status?
  [s]
  (= s db-encoding-non-reactive))

(defn terminal-status?
  [s]
  (contains?
    #{db-encoding-negative, db-encoding-reactive, db-encoding-non-reactive}
    s))


(defn configure-logging
  "Configures the logging for ctest. Log level and log file can be specified in the configuration."
  [{:keys [log-level, log-file] :as config}]
  (let [props (doto (System/getProperties)
                (.setProperty "log4j.rootLogger" (format "%s, file" (-> log-level name str/upper-case)))
                (.setProperty "log4j.appender.file" "org.apache.log4j.RollingFileAppender")
                (.setProperty "log4j.appender.file.File" (str log-file))
                (.setProperty "log4j.appender.file.MaxFileSize" "4MB")
                (.setProperty "log4j.appender.file.MaxBackupIndex" "5")
                (.setProperty "log4j.appender.file.layout" "org.apache.log4j.PatternLayout")
                (.setProperty "log4j.appender.file.layout.ConversionPattern" "%d{yyyy.MM.dd HH:mm:ss} %5p %c: %m%n")
                ; jetty is too chatty
                (.setProperty "log4j.logger.org.eclipse.jetty" "INFO"))]
    (PropertyConfigurator/configure props))
  nil)


(defn read-config
  [config-file]
  (try
    ; we trust anyone with access to the config file, so just load it
    (load-file config-file)
    (catch Throwable t
      (let [log-file "startup-errors.log"]
        (binding [log/*force* :direct]
          (configure-logging {:log-level :info, :log-file log-file})
          (log/errorf "Error when reading config file \"%s\":\n%s"
            config-file
            (with-out-str (st/print-cause-trace t)))
          (common/exit 2
            (format "Error when reading config file \"%s\": \"%s\"\nFor details see \"%s\"."
              config-file
              (.getMessage t)
              log-file)))))))


(defn db-exists?
  [config-filename, db-filename]
  (if (.exists (io/file db-filename))
    true
    (let [msg (format
                (str
                  "The database file \"%s\" does not exist!\n"
                  "You have the following two options to fix that:\n"
                  "(1) Fix the path to the existing database file in the configuration \"%s\".\n"
                  "(2) Rerun the ctest initialisation.")
                (-> db-filename io/file .getAbsolutePath),
                (-> config-filename io/file .getAbsolutePath))]
      (println msg)
      (log/error "Startup" msg)
      false)))


(defn read+setup-config
  [config-file]
  (let [{:keys [data-base-name] :as config} (read-config config-file)]
    (update-config config)
    (if-let [errors (check-config)]
      (do
        (println
         (format "The configuration file \"%s\" contains the following errors:\n  %s"
           config-file
           (str/join "\n  " errors)))
        false)
      (do
        (configure-logging config)
        (update-db-name data-base-name)
        (db-exists? config-file, data-base-name)))))
